crates/red_knot/src/types/infer.rs

```diff
@@ -16,21 +16,30 @@ use crate::{FileId, Name};
 
 // FIXME: Figure out proper dead-lock free synchronisation now that this takes `&db` instead of `&mut db`.
 #[tracing::instrument(level = "trace", skip(db))]
-pub fn infer_symbol_type(db: &dyn SemanticDb, symbol: GlobalSymbolId) -> QueryResult<Type> {
+pub fn infer_symbol_public_type(db: &dyn SemanticDb, symbol: GlobalSymbolId) -> QueryResult<Type> {
     let symbols = symbol_table(db, symbol.file_id)?;
     let defs = symbols.definitions(symbol.symbol_id);
     let jar: &SemanticJar = db.jar()?;
 
-    if let Some(ty) = jar.type_store.get_cached_symbol_type(symbol) {
+    if let Some(ty) = jar.type_store.get_cached_symbol_public_type(symbol) {
         return Ok(ty);
     }
 
-    // TODO handle multiple defs, conditional defs...
-    assert_eq!(defs.len(), 1);
-
-    let ty = infer_definition_type(db, symbol, defs[0].clone())?;
+    // The public type of a symbol is the union of all of its definitions. This is the most
+    // cautious/sound approach, though it can lead to a broader-than-desired type in a case like
+    // `x = 1; x = str(x)`, where the first definition of `x` can never be visible. TODO prune
+    // definitions that we can prove can't be visible.
+    let tys = defs
+        .iter()
+        .map(|def| infer_definition_type(db, symbol, def.clone()))
+        .collect::<QueryResult<Vec<Type>>>()?;
+    let ty = match tys.len() {
```

Comment:
We could possibly get away here without collecting in the most common case (in which case my suggestion below doesn't work. 

```rust
let mut tys = defs
        .iter()
        .map(|def| infer_definition_type(db, symbol, def.clone())).peekable();

if let Some(first) = tys.next() {
	if tys.peek().is_some() {
		Type::Union(jar.type_store.add_union(symbo.file_id, std::iter::chain([first], tys).collect()));
	} else {
		first
  }
} else {
	Type::Unkown
}
```

Reply:
This is really nice, thank you! With a few tweaks to deal with the QueryResult (and using `Iterator::chain` not `std::iter::chain`), I got it to work.